lib/Dialect/FIRRTL/Transforms/DropConst.cpp

```diff
@@ -0,0 +1,128 @@
+//===- DropConst.cpp - Check and remove const types -------------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This file defines the DropConst pass.
+//
+//===----------------------------------------------------------------------===//
+
+#include "PassDetails.h"
+#include "circt/Dialect/FIRRTL/FIRRTLOps.h"
+#include "circt/Dialect/FIRRTL/FIRRTLTypes.h"
+#include "circt/Dialect/FIRRTL/FIRRTLUtils.h"
+#include "circt/Dialect/FIRRTL/FIRRTLVisitors.h"
+#include "circt/Dialect/FIRRTL/Passes.h"
+#include "llvm/Support/SaveAndRestore.h"
+
+#define DEBUG_TYPE "drop-const"
+
+using namespace circt;
+using namespace firrtl;
+
+// NOLINTBEGIN(misc-no-recursion)
+/// Return this type with a 'const' modifiers dropped
+static FIRRTLBaseType getAllConstDroppedType(FIRRTLBaseType type) {
```

Comment:
consider renaming `getAllConstDroppedType` to `convertType`, and finding some other name for `convertType`.  You could introduce a helper:
```c++
static void updateType(Value &value) {
  auto type = convertType(value.getType());
  if (value.getType() != type)
    value.setType(type);
}
``` 

and i don't know if it would work:
```c++
llvm::for_each(op->getResults(), &updateType);
```

Reply:
I'm realizing now that `SubAccessOp` should infer an all const dropped type (instead of dropping only the top const like it does now). In that case `getAllConstDroppedType` should probably go back on the types.

Reply:
Hmm I have to think about it.  If a subaccess is used as a source, (`o <= c[x]`), then if the index is const, it should be fine to keep the constness (not just drop the outer const, right?); and if the index is not const then it should infer an "all const dropped" result.

If a subaccess is used as a sink (`c[x] <= i`), then we need ensure that .. uh .. lets just look at what it lowers to:
```
when (x == 0)
  c[0] <= i
else when (x == 0)
  c[1] <= i
else cetera
```
I guess this means we need to ensure that the index is const if it leads to a constant being written to, in which case you can probably use the same type inference rules (keep the constness when the index is constant).

Unless, there was already a discussion that subaccess always result in all non-dropped const results?

Reply:
Yeah, the inference rules should be the same regardless of usage as source or sink. This is just another place where the logic wasn't updated after the change from const being determined by the leaves to const being determined by the outer type. I bet there are a few places in my next pr that I need to update as well.

Reply:
Updated